src/topology/ping_pong.rs

```diff
@@ -0,0 +1,963 @@
+                    prep_share: prep_share.get_encoded(),
+                },
+            )
+        })
+        .map_err(PingPongError::VdafPrepareInit)
+    }
+
+    fn helper_initialized(
+        &self,
+        verify_key: &[u8; VERIFY_KEY_SIZE],
+        agg_param: &Self::AggregationParam,
+        nonce: &[u8; NONCE_SIZE],
+        public_share: &Self::PublicShare,
+        input_share: &Self::InputShare,
+        inbound: &PingPongMessage,
+    ) -> Result<Self::Transition, PingPongError> {
+        let (prep_state, prep_share) = self
+            .prepare_init(
+                verify_key,
+                /* Helper */ 1,
+                agg_param,
+                nonce,
+                public_share,
+                input_share,
+            )
+            .map_err(PingPongError::VdafPrepareInit)?;
+
+        let inbound_prep_share = if let PingPongMessage::Initialize { prep_share } = inbound {
+            Self::PrepareShare::get_decoded_with_param(&prep_state, prep_share)
+                .map_err(PingPongError::CodecPrepShare)?
+        } else {
+            return Err(PingPongError::PeerMessageStateMismatch {
+                found: inbound.variant(),
+                expected: "initialize",
+            });
+        };
+
+        let current_prepare_message = self
+            .prepare_preprocess([inbound_prep_share, prep_share])
+            .map_err(PingPongError::VdafPreparePreprocess)?;
+
+        Ok(PingPongTransition {
+            previous_prepare_state: prep_state,
+            current_prepare_message,
+        })
+    }
+
+    fn leader_continued(
+        &self,
+        leader_state: Self::State,
+        inbound: &PingPongMessage,
+    ) -> Result<Self::ContinuedValue, PingPongError> {
+        self.continued(true, leader_state, inbound)
+    }
+
+    fn helper_continued(
+        &self,
+        helper_state: Self::State,
+        inbound: &PingPongMessage,
+    ) -> Result<Self::ContinuedValue, PingPongError> {
+        self.continued(false, helper_state, inbound)
+    }
+}
+
+impl<const VERIFY_KEY_SIZE: usize, const NONCE_SIZE: usize, A>
+    PingPongTopologyPrivate<VERIFY_KEY_SIZE, NONCE_SIZE> for A
+where
+    A: Aggregator<VERIFY_KEY_SIZE, NONCE_SIZE>,
+{
+    fn continued(
+        &self,
+        is_leader: bool,
+        host_state: Self::State,
+        inbound: &PingPongMessage,
+    ) -> Result<Self::ContinuedValue, PingPongError> {
+        let host_prep_state = if let PingPongState::Continued(state) = host_state {
+            state
+        } else {
+            return Err(PingPongError::HostStateMismatch {
+                found: "finished",
+                expected: "continued",
+            });
+        };
+
+        let (prep_msg, next_peer_prep_share) = match inbound {
+            PingPongMessage::Initialize { .. } => {
+                return Err(PingPongError::PeerMessageStateMismatch {
+                    found: inbound.variant(),
+                    expected: "continued",
```

Comment:
nit: "continued" is a state, and "continue" is a message
```suggestion
                    expected: "continue",
```